python/tskit/metadata.py

```diff
@@ -0,0 +1,214 @@
+        return [replace_root_refs(j) for j in obj]
+    elif type(obj) == dict:
+        ret = {k: replace_root_refs(v) for k, v in obj.items()}
+        if ret.get("$ref") == "#":
+            ret["$ref"] = "#/definitions/root"
+        return ret
+    else:
+        return obj
+
+
+class TSKITMetadataSchemaValidator(jsonschema.validators.Draft7Validator):
+    """
+    Our schema is the Draft7Validator schema with added codec information.
+    """
+
+    META_SCHEMA: dict = copy.deepcopy(jsonschema.validators.Draft7Validator.META_SCHEMA)
+    # We need a top-level only required property so we need to rewite any reference
+    # to the top-level schema to a copy in a definition.
+    META_SCHEMA = replace_root_refs(META_SCHEMA)
+    META_SCHEMA["definitions"]["root"] = copy.deepcopy(META_SCHEMA)
+    META_SCHEMA["codec"] = {"type": "string"}
+    META_SCHEMA["required"] = ["codec"]
+
+
+class AbstractMetadataCodec(metaclass=abc.ABCMeta):
+    """
+    Superclass of all MetadataCodecs.
+    """
+
+    def __init__(self, schema: dict) -> None:
+        raise NotImplementedError  # pragma: no cover
+
+    @abc.abstractmethod
+    def encode(self, obj: Any) -> bytes:
+        raise NotImplementedError  # pragma: no cover
+
+    @abc.abstractmethod
+    def decode(self, encoded: bytes) -> Any:
+        raise NotImplementedError  # pragma: no cover
+
+
+codec_registry = {}
+
+
+def register_codec(codec_cls: Type[AbstractMetadataCodec], codec_id: str) -> None:
+    """
+    Register a metadata codec class.
+    This function maintains a mapping from metadata codec identifiers used in schemas
+    to codec classes. When a codec class is registered, it will replace any class
+    previously registered under the same codec identifier, if present.
+    See :ref:`sec_tutorial_metadata_custom_codec` for example usage.
+
+    :param str codec_id: String to use to refer to the codec in the schema.
+    """
+    codec_registry[codec_id] = codec_cls
+
+
+class JSONCodec(AbstractMetadataCodec):
+    def __init__(self, schema: dict) -> None:
+        pass
+
+    def encode(self, obj: Any) -> bytes:
+        return json.dumps(obj).encode()
+
+    def decode(self, encoded: bytes) -> Any:
+        return json.loads(encoded.decode())
+
+
+register_codec(JSONCodec, "json")
+
+
+class NOOPCodec(AbstractMetadataCodec):
+    def __init__(self, schema: dict) -> None:
+        pass
+
+    def encode(self, data: bytes) -> bytes:
+        return data
+
+    def decode(self, data: bytes) -> bytes:
+        return data
+
+
+def validate_bytes(data: Optional[bytes]) -> None:
+    if data is not None and not isinstance(data, bytes):
+        raise TypeError(
+            f"If no encoding is set metadata should be bytes, found {type(bytes)}"
+        )
+
+
+class MetadataSchema:
+    """
+    Class for validating, encoding and decoding metadata.
+
+    :param dict schema: A dict containing a valid JSONSchema object.
+        A good guide to creating a schema is at
```

Comment:
These two sentences don't really belong here, can delete as it's covered elsewhere.

Reply:
Fixed in https://github.com/tskit-dev/tskit/pull/543/commits/14b67649b0a1442f9b95ba53aa0e23fdf2053fe3